rekall/plugins/addrspaces/pmem.py

```diff
@@ -0,0 +1,341 @@
+    =======
+
+    IOError: If ioctl fails for whatever reason.
+    AssertionError: If you're using this code to hack an alien mothership
+                    and their EFI uses a different descriptor size than 0x30
+                    this will blow up.
+    """
+    mmap_size = pmem_get_mmap_size(fd)
+    mmap_desc_size = pmem_get_mmap_desc_size(fd)
+    mmap = ctypes.create_string_buffer(mmap_size * mmap_desc_size)
+
+    err = fcntl.ioctl(fd, PMEM_IOCTL_GET_MMAP, ctypes.pointer(mmap), True)
+    if err:
+        raise IOError("Error (%d) filling mmap buffer." % err)
+
+    assert(mmap_desc_size == 0x30,
+           ("EFI reports descriptor size of 0x%x. This code only knows how "
+            "to handle descriptors 0x30 bytes in length.") % mmap_desc_size)
+
+    return mmap, mmap_size, mmap_desc_size
+
+
+def pmem_parse_mmap(fd):
+    """Retrieve and parse the physical memory map from the Pmem driver.
+
+    Yields: tuples of (start, number of pages, type)
+    """
+    mmap, size, desc_size = pmem_get_mmap(fd)
+    session = rekall_session.Session()
+    buffer_as = addrspace.BufferAddressSpace(data=mmap.raw, session=session)
+    session.SetParameter("default_address_space", buffer_as)
+    profile = EFIProfile(session=session)
+
+    for offset in xrange(0, size, desc_size):
+        descriptor = profile.EFI_MEMORY_DESCRIPTOR(offset=offset)
+        yield (descriptor.PhysicalStart,
+               descriptor.NumberOfPages,
+               descriptor.Type)
+
+# See http://wiki.phoenix.com/wiki/index.php/EFI_MEMORY_TYPE for list of
+# segment types that become conventional memory after ExitBootServices()
+# is sent to EFI.
+EFI_SEGMENTS_SAFETY = {
+    "EfiReservedMemoryType": "",
+    "EfiLoaderCode": "rw",  # General use.
+    "EfiLoaderData": "rw",  # General use.
+    "EfiBootServicesCode": "rw",  # General use.
+    "EfiBootServicesData": "rw",  # General use.
+    "EfiRuntimeServicesCode": "r",  # Memory to be preserved.
+    "EfiRuntimeServicesData": "r",  # Memory to be preserved.
+    "EfiConventionalMemory": "r",  # General use.
+    "EfiUnusableMemory": "",  # (Hardware) errors - don't use.
+    "EfiACPIReclaimMemory": "rw",  # General use after ACPI enabled.
+    "EfiACPIMemoryNVS": "r",  # Memory to be preserved.
+    "EfiMemoryMappedIO": "",  # ACPI tables.
+    "EfiMemoryMappedIOPortSpace": "",  # ACPI tables.
+    "EfiPalCode": "r",  # OS-dependent. Largely read-only.
+    "EfiMaxMemoryType": "rw",  # No idea (adamsh). Looks like general use?
+}
+
+
+def efi_type_writable(efi_type):
+    return "w" in EFI_SEGMENTS_SAFETY[str(efi_type)]
+
+
+def efi_type_readable(efi_type):
+    return "r" in EFI_SEGMENTS_SAFETY[str(efi_type)]
+
+
+# Adapted from http://wiki.phoenix.com/wiki/index.php/EFI_MEMORY_DESCRIPTOR
+#
+# Every kernel that supports EFI will have a similar struct, but called
+# something else. For example, on XNU this is called EfiMemoryRange, but
+# XNU's definition doesn't account for the trailing 8-byte padding.
```

Comment:
you actually dont need to add a padding field here at all. As long as you have the struct's size defined thats enough.